lib/collection/src/lib.rs

```diff
@@ -886,6 +898,99 @@ impl Collection {
+        }
+        Ok(snapshot_path)
+    }
+
+    pub async fn create_snapshot(&self) -> CollectionResult<SnapshotDescription> {
+        let snapshot_name = format!(
+            "{}-{}.snapshot",
+            self.name(),
+            chrono::Utc::now().format("%Y-%m-%d-%H-%M-%S")
+        );
+        let snapshot_path = self.snapshots_path.join(snapshot_name);
+
+        let snapshot_path_with_tmp_extension = snapshot_path.clone().with_extension("tmp");
+        let snapshot_path_with_arc_extension = snapshot_path.clone().with_extension("arc");
+        create_dir_all(&snapshot_path_with_tmp_extension).await?;
+
+        // Create snapshot of each shard
+        for (shard_id, shard) in self.shards.iter() {
+            let shard_snapshot_path =
+                snapshot_path_with_tmp_extension.join(format!("{}", shard_id));
+            create_dir_all(&shard_snapshot_path).await?;
+            match shard {
+                Shard::Local(local_shard) => {
+                    local_shard.create_snapshot(&shard_snapshot_path).await?;
+                }
+                Shard::Remote(remote_shard) => {
+                    // copy shard directory to snapshot directory
+                    remote_shard.create_snapshot(&shard_snapshot_path).await?;
+                }
+            }
+        }
+
+        CollectionVersion::save(&snapshot_path_with_tmp_extension)?;
+        self.config
+            .read()
+            .await
+            .save(&snapshot_path_with_tmp_extension)?;
+
+        // have to use std here, cause TarBuilder is not async
+        let file = std::fs::File::create(&snapshot_path_with_arc_extension)?;
+        let mut builder = TarBuilder::new(file);
+        // archive recursively collection directory `snapshot_path_with_arc_extension` into `snapshot_path`
+        builder.append_dir_all(".", &snapshot_path_with_tmp_extension)?;
+        builder.finish()?;
+
+        // remove temporary snapshot directory
+        remove_dir_all(&snapshot_path_with_tmp_extension).await?;
+
+        // move snapshot to permanent location
+        rename(&snapshot_path_with_arc_extension, &snapshot_path).await?;
+
+        get_snapshot_description(&snapshot_path).await
+    }
+
+    pub fn restore_snapshot(snapshot_path: &Path, target_dir: &Path) -> CollectionResult<()> {
+        // decompress archive
+        let archive_file = std::fs::File::open(snapshot_path).unwrap();
+        let mut ar = tar::Archive::new(archive_file);
+        ar.unpack(target_dir)?;
+
+        let config = CollectionConfig::load(target_dir).unwrap_or_else(|err| {
+            panic!(
+                "Can't read collection config due to {}\nat {}",
+                err,
+                target_dir.to_str().unwrap()
+            )
+        });
+        let configured_shards = config.params.shard_number.get();
+
+        for shard_id in 0..configured_shards {
+            let shard_path = shard_path(target_dir, shard_id);
+            if shard_path.exists() {
```

Comment:
Why do we need to check for existence here? Is it possible to have a shard missing?